ServiceTier/Services/LogInService.cs

```diff
@@ -33,39 +36,60 @@ public string CreateToken(LoginDTO model)
                 throw new UserNotFoundException();
             }
 
-            if(!userManager.UserManager.CheckPasswordAsync(user, model.Password).Result)
+            if (!userManager.UserManager.CheckPasswordAsync(user, model.Password).Result)
             {
                 throw new IncorectLoginException();
             }
 
-            var token = GenerateToken(user);
+            var token = JWTManager.GenerateToken(user);
 
             return token;
         }
 
-        private static string GenerateToken(User user)
+        public string CreateTokenWithGoogle(SocialAuthDTO accessToken)
         {
-            var claims = new[]
-            {
-                new Claim(Constants.ClaimUserId, user.Id.ToString()),
-                new Claim(JwtRegisteredClaimNames.Sub, user.Email),
-                new Claim(JwtRegisteredClaimNames.Jti, Guid.NewGuid().ToString())
-            };
+            string userInfoResponse = Client.GetStringAsync($"{Constants.GoogleResponsePath}{accessToken.AccessToken}").Result;
+
+            var userInfo = JsonConvert.DeserializeObject<GoogleAuthUserData>(userInfoResponse);
+            User user = userManager.UserManager.FindByEmailAsync(userInfo.Email).Result;
 
-            var signingKey = new SymmetricSecurityKey(Encoding.UTF8.GetBytes(IoCContainer.Configuration["JWTSecretKey"]));
+            if (user == null)
+            {
+                user = userInfo.ToUser();
+                var result = userManager.UserManager.CreateAsync(user, Convert.ToBase64String(Guid.NewGuid().ToByteArray()).Substring(0, 8)).Result;
 
-            var token = new JwtSecurityToken(
-                expires: DateTime.UtcNow.AddHours(Constants.Validity),
-                claims: claims,
-                signingCredentials: new SigningCredentials(signingKey, SecurityAlgorithms.HmacSha256),
-                issuer: "Tokens:Issuer",
-                audience: "Tokens:Audience"
-                );
+                if (!result.Succeeded)
+                {
+                    throw new UserNotAddedException();
+                }
+            }
 
-            var tokenString = new JwtSecurityTokenHandler().WriteToken(token);
+            string jwt = JWTManager.GenerateToken(user);
 
-            return JsonConvert.SerializeObject(new { access_token = tokenString });
+            return jwt;
         }
 
+        public string CreateTokenWithFacebook(SocialAuthDTO accessToken)
```

Comment:
You right, parts of code seem duplicates, but actually, this code parts are similar, but not same. For example, when is invoked extension method toUser we pass different parameters type (FacebookAuthUserData and GoogleAuthUserData), and I have decided don't extract this small part same code to separate method.

Reply:
Yes, I can see they are not completely the same. But, having different params doesn't justify duplicated code. So again - extract similar code to separate method.

Reply:
ok, extract similar code

Reply:
In this two method the same code is several lines (circled). I don`t know if it needs to move to a separate method. I can`t take more code because variable userInfo has different type: FacebookAuthUserData and GoogleAuthUserData in this two method, that is why when I will find user by email it will need put in IF block, also when user don`t exist and I invoke extension method toUser I pass two different types in this two method, that is why I need put in If block, if put it in separate method.
![image](https://user-images.githubusercontent.com/37620894/50979824-00da4a80-1500-11e9-96f6-7cab65c7d330.png)